misc/codepoint_width.c

```diff
@@ -0,0 +1,758 @@
+    0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x40, 0x42,
+    0x40, 0x40, 0x02, 0x02, 0x42, 0x02, 0x44, 0x42,
+    0x44, 0x42, 0x02, 0x40, 0x40, 0x40, 0x40, 0x02,
+    0x02, 0x42, 0x42, 0x42, 0x42, 0x02, 0x40, 0x40,
+    0x40, 0x42, 0x40, 0x40, 0x40, 0x02, 0x42, 0x44,
+    0x02, 0x02, 0x02, 0x02, 0x40, 0x40, 0x40, 0x40,
+    0x44, 0x44, 0x44, 0x44, 0x02, 0x02, 0x02, 0x02,
+    0x02, 0x02, 0x40, 0x42, 0x02, 0x02, 0x02, 0x02,
+    0x02, 0x02, 0x42, 0x02, 0x02, 0x42, 0x02, 0x02,
+    0x40, 0x02, 0x02, 0x02, 0x02, 0x02, 0x02, 0x40,
+    0x40, 0x02, 0x40, 0x02, 0x02, 0x40, 0x02, 0x02,
+    0x02, 0x02, 0x02, 0x02, 0x44, 0x02, 0x40, 0x40,
+    0x42, 0x42, 0x42, 0x42, 0x42, 0x40, 0x02, 0x02,
+    0x40, 0x42, 0x42, 0x02, 0x42, 0x02, 0x40, 0x40,
+    0x40, 0x02, 0x02, 0x42, 0x42, 0x40, 0x02, 0x02,
+    0x44, 0x42, 0x40, 0x40, 0x40, 0x40, 0x02, 0x02,
+    0x02, 0x40, 0x40, 0x40, 0x42, 0x42, 0x02, 0x42,
+    0x02, 0x40, 0x40, 0x40, 0x40, 0x40, 0x46, 0x40,
+    0x40, 0x40, 0x46, 0x46, 0x46, 0x40, 0x40, 0x40,
+    0x40, 0x40, 0x80, 0x80, 0x80, 0x80, 0x02, 0x40,
+    0x40, 0x40, 0x82, 0x82, 0x40, 0x40, 0x40, 0x40,
+    0x40, 0x40, 0x80, 0x80, 0x80, 0x80, 0x40, 0x80,
+    0x80, 0x40, 0x40, 0x80, 0x40, 0x40, 0x40, 0x40,
+    0x40, 0x80, 0x80, 0x80, 0x40, 0x40, 0x80, 0x40,
+    0x40, 0x01, 0x01, 0x01, 0x01, 0x40, 0x40, 0x40,
+    0x40, 0x42, 0x42, 0x42, 0x01, 0x01, 0x01, 0x01,
+    0x01, 0x02, 0x02, 0x02, 0x02, 0x02, 0x40, 0x40,
+    0x02, 0x02, 0x02, 0x02, 0x02, 0x4c, 0x4c, 0x4c,
+    0x4c, 0x8c, 0x4c, 0x4c, 0x4c, 0xc0, 0xc0, 0xc0,
+    0x40, 0x40, 0x4c, 0x4c, 0x4c, 0xc0, 0xc0, 0xc0,
+    0xc0, 0xc0, 0xc0, 0x40, 0x4c, 0xc0, 0xc0, 0x40,
+    0x40, 0x4c, 0x4c, 0x4c, 0x4c, 0xcc, 0xcc, 0xc0,
+    0xc0, 0xc0, 0xc0, 0xc0, 0xc0, 0xcc, 0xcc, 0xc0,
+    0xc0, 0xc0, 0xc0, 0xc0, 0xc0, 0x8c, 0xc0, 0x8c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0xc0, 0xc0,
+    0xc0, 0xc0, 0xc0, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c,
+    0x4c, 0x43, 0x43, 0x43, 0x43, 0x43, 0x43, 0x43,
+    0x43, 0x80, 0x8c, 0x8c, 0x4c, 0x4c, 0x4c, 0x4c,
+    0x4c, 0x80, 0x80, 0x8c, 0x80, 0x80, 0x80, 0x80,
+    0x80, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x80,
+    0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x8c,
+    0x8c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x4c, 0x8c,
+    0x8c, 0x8c, 0x4c, 0x4c, 0x4c, 0x4c, 0x8c, 0x4c,
+    0x4c, 0x4c, 0x8c, 0x4c, 0x4c, 0x4c, 0x8c, 0x8c,
+    0x8c, 0x82, 0x82, 0x82, 0x82, 0x82, 0x8c, 0x4c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x4c, 0x4c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x40, 0x40,
+    0x4c, 0x4c, 0x4c, 0x8c, 0x8c, 0x8c, 0x8c, 0x4c,
+    0x40, 0x40, 0x40, 0x40, 0x4c, 0x4c, 0x4c, 0x4c,
+    0x4c, 0x4c, 0x40, 0x40, 0x40, 0x40, 0x8c, 0x8c,
+    0x8c, 0x8c, 0x40, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c,
+    0x8c, 0x40, 0x8c, 0x41, 0x01, 0x41, 0x41, 0x41,
+    0x41, 0x41, 0x41, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0,
+    0xc0, 0x40, 0x40,
+};
+static const uint32_t s_joinRules[2][16] = {
+    {
+        0b00000011110011111111111111001111,
```

Comment:
mpv is not exactly free from extensions and this works fine on all compilers that we support.

Reply:
I think it the case above it actually were throwing warnings and failing to compile. Not saying using non standard extensions is good, but we need to be pragmatic here. I really want to avoid vendoring generator / rewriting / maintaining it. As you may have noticed. Let say it another way, I won't do that myself, I won't oppose if anyone else wants to do it though.

Changing this from binary representation is small change, but once we start doing it the process of updating is no longer "easy", because require modification of generation code and so on.

Reply:
> Right now the generation process requires some manual fiddling anyway. I think it's a good idea to add a script in TOOLS which automates the whole process providing that .NET is installed. Once that's on the table the replacement can be done in the script.

I think we're focusing too much on automating something that's only updated every 1-2 years at best. While it would be ideal to fully automate and just watch the code being written, it's not that simple.

Maybe my reasoning isn't clear, so let me explain. I'm not interested in maintaining this generator in the mpv tree because it’s not necessary. I understand you're not suggesting that exactly, but having a script that does generation and adjustment automatically, is not easy either. Sure, s/constexpr/const/ is easy, but replacing auto is not. We'd need to build an AST to figure out the correct type for auto (or ask an LLM). It’s not worth the effort. So, we're back to either creating our own generator or modifying the upstream one, which I don't think is needed.

The issue is that in 1-2 years, when this script is used again, the upstream version will likely be updated, rendering any static patches we create now useless. Since updates are so infrequent, we can afford to revisit this topic in a couple of years and update the version and review the changes that may have happened upstream.

I'm happy that there's an easy way to generate the code. It might not be a perfect drop-in replacement, but the "manual fiddling" only involves changing a dozen of lines. We can't predict what will happen in the future... there could be a completely different algorithm needed.

It's not that I don't want to automate this. I do I like automation, as you probably know if you've followed my CI work. I just don't know how to do it without maintaining our own script, which I’d rather avoid for something used so rarely.

That said, any suggestions or code examples on how to achieve this are very welcome.

But if this is a blocker for the PR, so be it. It wouldn’t be the first time a nice feature is blocked over a mundane issue.

> For now it's OK, but it will become a problem if we bump the warning level to 3 which enables -Wpedantic.

Don't raise the bar for this PR. It doesn't matter at this point. There are still unresolved warnings at the current level that are being suppressed, and using pedantic would generate even more. I'm all for resolving warnings, but if it impacts other aspects, it's not always beneficial to force fixing them.

Reply:
> I think we're focusing too much on automating something that's only updated every 1-2 years at best

The last two updates were few months apart (15.1 and 16.0).

Reply:
> The last two updates were few months apart (15.1 and 16.0).

12 months apart to be exact.

Reply:
> If it's the case that the frequency of the updates doesn't justify the efforts for automation (which is my initial thought), then what's the issue with manually converting the numeric format for these infrequent situations? Compared to potential incompatible upstream changes that require more manual fix, this effort seems minuscule.

Fair point. I was focused on minimizing changes as much as possible, but I can agree to do this change too. Changed now to base 16.

Reply:
> Downgrading the algorithm to bisection without clusters seems unnecessary.

I think you miss that the Grapheme clusters code is mostly no-op. wcwidth alone can handle Grapheme clusters very well, with very few exceptions.

These exceptions are very few codepoints which do have a width of their own (e.g. 2), but their main usage is combined with others as modifiers, typically/exclussively with emojis, at which case their width (in a series following an emoji) becomes 0.

If the wcwidth implementation assigns them a width of 0 (which for instance the code at #14466 does), then this covers what Grapheme clusters analysis does. It does leave out that it will not detect of these typically-modifiers on its own, but that's a very very rare use case in practice.

For instance emoji skin tone modifiers on their own would look like a sqare in a specific skin tone, but that doesn't really happen with normal content, because mostly it modifies an emoji, so if wcwidth returns 0 for these modifiers, then you don't need the grapheme clusters analysis for these cases.

Not completely perfect, but effectively perfect.

And the added value is that you need a one page script to generate it instead of dot-net sdk, and can tweak it however you see fit, because it's a tiny generator.

The balance is between cost and value. And when the value is effectively the same, it comes down to cost.

Reply:
> And when the value is effectively the same...

I don't know if you're trying to troll me or if something is genuinely wrong here. I've explained multiple times why the value is not the same. Grapheme clusters are not supported, and the bisection method is naive and not optimal for this task. You can't ignore all the facts in the trade-off analysis and claim that solution A is the same as solution B just because they work the same "90%" of the time.

You are spending whole paragraphs of text explaining why inferior solution is "almost the same", "not completely perfect", what is the reason for that? Why current solution cannot be used?

> The balance is between cost and value.

Exactly. The cost of updating this code every 1-2 years is almost zero. That's why I don't see why using inferior solution would be preferable? Maintaining even simple script would have higher cost than just running existing solution. I still didn't see any argument why using this generator is "costly" or "hard" except initial misunderstanding about .NET.

Reply:
Nevermind. Please ignore my comment. I'm leaving it here only for posterity.

Reply:
As I mentioned earlier, I'm open to other solutions. If you submit a PR with code that is better by your standards, I'd be happy to replace the current implementation. However, at this point, I don't see any major issues with the current approach that would prevent it from being used.